File: src/api/org/openmrs/api/UserService.java
/**
 * The contents of this file are subject to the OpenMRS Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
 */
package org.openmrs.api;

import java.util.List;

import org.openmrs.Privilege;
import org.openmrs.Role;
import org.openmrs.User;
import org.openmrs.annotation.Authorized;
import org.openmrs.annotation.Logging;
import org.openmrs.util.OpenmrsConstants;
import org.springframework.transaction.annotation.Transactional;

/**
 * Contains methods pertaining to Users in the system Use:<br/>
 * 
 * <pre>
 *   List<User> users = Context.getUserService().getAllUsers();
 * </pre>
 * 
 * @see org.openmrs.api.context.Context
 */
@Transactional
public interface UserService extends OpenmrsService {
	
	/**
	 * Saves a user to the database.
	 * 
	 * @param user
	 * @param password
	 * @returns a User object
	 * @throws APIException
	 */
	@Authorized( { OpenmrsConstants.PRIV_ADD_USERS, OpenmrsConstants.PRIV_EDIT_USERS })
	@Logging(ignoredArgumentIndexes = { 1 })
	public User saveUser(User user, String password) throws APIException;
	
	/**
	 * @see {@link #saveUser(User, String)}
	 * @deprecated replaced by {@link #saveUser(User, String)}
	 */
	@Authorized( { OpenmrsConstants.PRIV_ADD_USERS })
	@Logging(ignoredArgumentIndexes = { 1 })
	public User createUser(User user, String password) throws APIException;
	
	/**
	 * Get user by internal user identifier.
	 * 
	 * @param userId internal identifier
	 * @return requested user
	 * @throws APIException
	 */
	@Transactional(readOnly = true)
	@Authorized( { OpenmrsConstants.PRIV_VIEW_USERS })
	public User getUser(Integer userId) throws APIException;
	
	/**
	 * Get user by username (user's login identifier)
	 * 
	 * @param username user's identifier used for authentication
	 * @return requested user
	 * @throws APIException
	 * @should get user by username
	 */
	@Transactional(readOnly = true)
	@Authorized( { OpenmrsConstants.PRIV_VIEW_USERS })
	public User getUserByUsername(String username) throws APIException;
	
	/**
	 * true/false if username or systemId is already in db in username or system_id columns
	 * 
	 * @param User to compare
	 * @return boolean
	 * @throws APIException
	 */
	@Transactional(readOnly = true)
	@Authorized( { OpenmrsConstants.PRIV_VIEW_USERS })
	public boolean hasDuplicateUsername(User user) throws APIException;
	
	/**
	 * Get users by role granted
	 * 
	 * @param Role role that the Users must have to be returned
	 * @return users with requested role
	 * @throws APIException
	 */
	@Transactional(readOnly = true)
	@Authorized( { OpenmrsConstants.PRIV_VIEW_USERS })
	public List<User> getUsersByRole(Role role) throws APIException;
	
	/**
	 * Save changes to given <code>user</code> to the database.
	 * 
	 * @param user
	 * @throws APIException
	 * @see {@link #saveUser(User, String)}
	 * @deprecated replaced by {@link #saveUser(User, String)}
	 */
	@Authorized( { OpenmrsConstants.PRIV_EDIT_USERS })
	public void updateUser(User user) throws APIException;
	
	/**
	 * Use should be UserService.saveUser(user.addRole(role))
	 * 
	 * @deprecated use {@link org.openmrs.User#addRole(Role)}
	 */
	@Authorized( { OpenmrsConstants.PRIV_EDIT_USERS })
	public void grantUserRole(User user, Role role) throws APIException;
	
	/**
	 * Use UserService.saveUser(user.removeRole(role))
	 * 
	 * @deprecated use {@link org.openmrs.User#removeRole(Role)}
	 */
	@Authorized( { OpenmrsConstants.PRIV_EDIT_USERS })
	public void revokeUserRole(User user, Role role) throws APIException;
	
	/**
	 * Mark user as voided (effectively deleting user without removing their data &mdash; since
	 * anything the user touched in the database will still have their internal identifier and point
	 * to the voided user for historical tracking purposes.
	 * 
	 * @param user
	 * @param reason
	 * @return the given user voided out
	 * @throws APIException
	 */
	@Authorized( { OpenmrsConstants.PRIV_EDIT_USERS })
	public User voidUser(User user, String reason) throws APIException;
	
	/**
	 * Clear voided flag for user (equivalent to an "undelete" or Lazarus Effect for user)
	 * 
	 * @param user
	 * @return the given user unvoided
	 * @throws APIException
	 */
	@Authorized( { OpenmrsConstants.PRIV_EDIT_USERS })
	public User unvoidUser(User user) throws APIException;
	
	/**
	 * @see #voidUser(User, String)
	 * @see #purgeUser(User)
	 * @deprecated use {@link #purgeUser(User)}
	 */
	@Authorized( { OpenmrsConstants.PRIV_DELETE_USERS })
	public void deleteUser(User user) throws APIException;
	
	/**
	 * Completely remove a location from the database (not reversible). This method delegates to
	 * #purgeLocation(location, boolean) method.
	 * 
	 * @param user the User to remove from the database.
	 */
	@Authorized( { OpenmrsConstants.PRIV_PURGE_USERS })
	public void purgeUser(User user) throws APIException;
	
	/**
	 * Completely remove a user from the database (not reversible). This is a delete from the
	 * database. This is included for troubleshooting and low-level system administration. Ideally,
	 * this method should <b>never</b> be called &mdash; <code>Users</code> should be
	 * <em>voided</em> and not <em>deleted</em> altogether (since many foreign key constraints
	 * depend on users, deleting a user would require deleting all traces, and any historical trail
	 * would be lost). This method only clears user roles and attempts to delete the user record. If
	 * the user has been included in any other parts of the database (through a foreign key), the
	 * attempt to delete the user will violate foreign key constraints and fail.
	 * 
	 * @param cascade <code>true</code> to delete associated content
	 */
	@Authorized( { OpenmrsConstants.PRIV_PURGE_USERS })
	public void purgeUser(User user, boolean cascade) throws APIException;
	
	/**
	 * @deprecated use {@link #getAllPrivileges()}
	 */
	@Transactional(readOnly = true)
	public List<Privilege> getPrivileges() throws APIException;
	
	/**
	 * Returns all privileges currently possible for any User
	 * 
	 * @return Global list of privileges
	 * @throws APIException
	 */
	@Transactional(readOnly = true)
	public List<Privilege> getAllPrivileges() throws APIException;
	
	/**
	 * @deprecated use {@link #getAllRoles()}
	 */
	@Transactional(readOnly = true)
	public List<Role> getRoles() throws APIException;
	
	/**
	 * Returns all roles currently possible for any User
	 * 
	 * @return Global list of roles
	 * @throws APIException
	 */
	@Transactional(readOnly = true)
	public List<Role> getAllRoles() throws APIException;
	
	/**
	 * @deprecated use {@link org.openmrs.Role#getInheritedRoles()}
	 */
	@Transactional(readOnly = true)
	public List<Role> getInheritingRoles(Role role) throws APIException;
	
	/**
	 * Save the given role in the database
	 * 
	 * @param Role to update
	 * @return the saved role
	 * @throws APIException
	 */
	@Authorized( { OpenmrsConstants.PRIV_MANAGE_ROLES })
	public Role saveRole(Role role) throws APIException;
	
	/**
	 * Complete remove a role from the database
	 * 
	 * @param Role to delete from the database
	 * @throws APIException
	 */
	@Authorized( { OpenmrsConstants.PRIV_PURGE_ROLES })
	public void purgeRole(Role role) throws APIException;
	
	/**
	 * Save the given privilege in the database
	 * 
	 * @param Privilege to update
	 * @return the saved privilege
	 * @throws APIException
	 */
	@Authorized( { OpenmrsConstants.PRIV_MANAGE_PRIVILEGES })
	public Privilege savePrivilege(Privilege privilege) throws APIException;
	
	/**
	 * Complete remove a privilege from the database
	 * 
	 * @param Privilege to delete
	 * @throws APIException
	 */
	@Authorized( { OpenmrsConstants.PRIV_PURGE_PRIVILEGES })
	public void purgePrivilege(Privilege privilege) throws APIException;
	
	/**
	 * Returns role object with given string role
	 * 
	 * @return Role
	 * @throws APIException
	 */
	@Transactional(readOnly = true)
	public Role getRole(String r) throws APIException;
	
	/**
	 * Returns Privilege in the system with given String privilege
	 * 
	 * @return Privilege
	 * @throws APIException
	 */
	@Transactional(readOnly = true)
	public Privilege getPrivilege(String p) throws APIException;
	
	/**
	 * @deprecated use {@link #getAllUsers()}
	 */
	@Transactional(readOnly = true)
	@Authorized( { OpenmrsConstants.PRIV_VIEW_USERS })
	public List<User> getUsers() throws APIException;
	
	/**
	 * Returns all users in the system
	 * 
	 * @return Global list of users
	 * @throws APIException
	 */
	@Transactional(readOnly = true)
	@Authorized( { OpenmrsConstants.PRIV_VIEW_USERS })
	public List<User> getAllUsers() throws APIException;
	
	/**
	 * Changes the <code>user<code>'s password
	 * ** Restricted to Super User access**
	 * 
	 * @param u user
	 * @param pw2 new password
	 * @throws APIException
	 */
	@Authorized( { OpenmrsConstants.PRIV_EDIT_USER_PASSWORDS })
	@Logging(ignoredArgumentIndexes = { 1 })
	public void changePassword(User u, String pw) throws APIException;
	
	/**
	 * Changes the current user's password.
	 * 
	 * @param pw current password
	 * @param pw2 new password
	 * @throws APIException
	 * @should matchOnCorrectlyHashedStoredPassword
	 * @should matchOnIncorrectlyHashedStoredPassword
	 */
	@Logging(ignoredArgumentIndexes = { 0, 1 })
	public void changePassword(String pw, String pw2) throws APIException;
	
	/**
	 * Changes the current user's password directly. This is most useful if migrating users from
	 * other systems and you want to retain the existing passwords. This method will simply save the
	 * passed hashed password and salt directly to the database.
	 * 
	 * @param user the user whose password you want to change
	 * @param hashedPassword - the <em>already hashed</em> password to store
	 * @param salt - the salt which should be used with this hashed password
	 * @throws APIException
	 * @should successfullySaveToTheDatabase
	 */
	@Authorized( { OpenmrsConstants.PRIV_EDIT_USER_PASSWORDS })
	public void changeHashedPassword(User user, String hashedPassword, String salt) throws APIException;
	
	/**
	 * Changes the current user's secret question and answer.
	 * 
	 * @param pw user's password
	 * @param question
	 * @param answer
	 * @throws APIException
	 * @should matchOnCorrectlyHashedStoredPassword
	 * @should matchOnIncorrectlyHashedStoredPassword
	 */
	@Logging(ignoreAllArgumentValues = true)
	public void changeQuestionAnswer(String pw, String q, String a) throws APIException;
	
	/**
	 * Compares <code>answer</code> against the <code>user</code>'s secret answer.
	 * 
	 * @param user
	 * @param answer
	 * @throws APIException
	 */
	@Transactional(readOnly = true)
	@Logging(ignoredArgumentIndexes = { 1 })
	public boolean isSecretAnswer(User u, String answer) throws APIException;
	
	/**
	 * Return a user if any part of the search matches first/last/system id and the user has one at
	 * least one of the given <code>roles</code> assigned to them
	 * 
	 * @param nameSearch
	 * @param roles
	 * @param includeVoided
	 * @return
	 */
	@Transactional(readOnly = true)
	@Authorized( { OpenmrsConstants.PRIV_VIEW_USERS })
	public List<User> getUsers(String nameSearch, List<Role> roles, boolean includeVoided) throws APIException;
	
	/**
	 * @deprecated use {@link #getUsers(String, List, boolean)}
	 */
	@Transactional(readOnly = true)
	@Authorized( { OpenmrsConstants.PRIV_VIEW_USERS })
	public List<User> findUsers(String name, List<String> roles, boolean includeVoided) throws APIException;
	
	/**
	 * @deprecated use {@link #getUsersByName(String, String, boolean)}
	 */
	@Transactional(readOnly = true)
	@Authorized( { OpenmrsConstants.PRIV_VIEW_USERS })
	public List<User> findUsers(String givenName, String familyName, boolean includeVoided) throws APIException;
	
	/**
	 * Search for a list of users by exact first name and last name.
	 * 
	 * @param givenName
	 * @param familyName
	 * @param includeVoided
	 * @return
	 */
	@Transactional(readOnly = true)
	@Authorized( { OpenmrsConstants.PRIV_VIEW_USERS })
	public List<User> getUsersByName(String givenName, String familyName, boolean includeVoided) throws APIException;
	
	/**
	 * @deprecated use {@link #getUsers(String, List, boolean)}
	 */
	@Transactional(readOnly = true)
	@Authorized( { OpenmrsConstants.PRIV_VIEW_USERS })
	public List<User> getAllUsers(List<Role> roles, boolean includeVoided) throws APIException;
	
	/**
	 * Adds the <code>key</code>/<code>value</code> pair to the given <code>user</code>.
	 * <b>Implementations of this method should handle privileges</b>
	 * 
	 * @param user
	 * @param key
	 * @param value
	 * @return the user that was passed in and added to
	 */
	public User setUserProperty(User user, String key, String value) throws APIException;
	
	/**
	 * Removes the property denoted by <code>key</code> from the <code>user</code>'s properties.
	 * <b>Implementations of this method should handle privileges</b>
	 * 
	 * @param user
	 * @param key
	 * @return the user that was passed in and removed from
	 */
	public User removeUserProperty(User user, String key) throws APIException;
	
	/**
	 * Get/generate/find the next system id to be doled out. Assume check digit /not/ applied in
	 * this method
	 * 
	 * @return new system id
	 */
	public String generateSystemId() throws APIException;
	
}

File: api/src/main/kotlin/org/openmrs/api/UserService.kt
/**
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */
package org.openmrs.api

import org.openmrs.Person
import org.openmrs.Privilege
import org.openmrs.Role
import org.openmrs.User
import org.openmrs.annotation.Authorized
import org.openmrs.annotation.Logging
import org.openmrs.notification.MessageException
import org.openmrs.util.PrivilegeConstants
import java.util.Locale

/**
 * Contains methods pertaining to Users in the system.
 *
 * Use:
 * ```
 * val users = Context.getUserService().getAllUsers()
 * ```
 *
 * @see org.openmrs.api.context.Context
 */
interface UserService : OpenmrsService {

    companion object {
        const val ADMIN_PASSWORD_LOCKED_PROPERTY = "admin_password_locked"
    }

    /**
     * Create user with given password.
     *
     * @param user the user to create
     * @param password the password for created user
     * @return created user
     * @throws APIException if creation fails
     */
    @Authorized(PrivilegeConstants.ADD_USERS)
    @Logging(ignoredArgumentIndexes = [1])
    @Throws(APIException::class)
    fun createUser(user: User, password: String): User

    /**
     * Change user password.
     *
     * @param user the user to update password
     * @param oldPassword the user password to update
     * @param newPassword the new user password
     * @throws APIException for not existing user and if old password is weak
     * @since 1.12
     */
    @Authorized(PrivilegeConstants.EDIT_USER_PASSWORDS)
    @Logging(ignoredArgumentIndexes = [1, 2])
    @Throws(APIException::class)
    fun changePassword(user: User, oldPassword: String?, newPassword: String)

    /**
     * Get user by internal user identifier.
     *
     * @param userId internal identifier
     * @return requested user
     * @throws APIException if retrieval fails
     */
    @Authorized(PrivilegeConstants.GET_USERS)
    @Throws(APIException::class)
    fun getUser(userId: Int?): User?

    /**
     * Get user by the given uuid.
     *
     * @param uuid the uuid to search for
     * @return user or null
     * @throws APIException if retrieval fails
     */
    @Authorized(PrivilegeConstants.GET_USERS)
    @Throws(APIException::class)
    fun getUserByUuid(uuid: String): User?

    /**
     * Get user by username (user's login identifier).
     *
     * @param username User's identifier used for authentication
     * @return requested user
     */
    @Authorized(PrivilegeConstants.GET_USERS)
    fun getUserByUsername(username: String): User?

    /**
     * Gets a user by username or email.
     *
     * @param usernameOrEmail User's email address or username
     * @return requested user or null if not found
     */
    @Authorized(PrivilegeConstants.GET_USERS)
    fun getUserByUsernameOrEmail(usernameOrEmail: String): User?

    /**
     * Gets a user with the specified activation key.
     *
     * @param activationKey User's activation key for password reset
     * @return requested User with associated activation key
     */
    @Authorized(PrivilegeConstants.GET_USERS)
    fun getUserByActivationKey(activationKey: String): User?

    /**
     * true/false if username or systemId is already in db in username or system_id columns.
     *
     * @param user User to compare
     * @return boolean
     * @throws APIException if check fails
     */
    @Authorized(PrivilegeConstants.GET_USERS)
    @Throws(APIException::class)
    fun hasDuplicateUsername(user: User): Boolean

    /**
     * Get users by role granted.
     *
     * @param role Role that the Users must have to be returned
     * @return users with requested role
     * @throws APIException if retrieval fails
     */
    @Authorized(PrivilegeConstants.GET_USERS)
    @Throws(APIException::class)
    fun getUsersByRole(role: Role): List<User>

    /**
     * Updates a given user in the database.
     *
     * @param user the user to save
     * @return the saved user
     * @throws APIException if saving fails
     */
    @Authorized(PrivilegeConstants.EDIT_USERS)
    @Throws(APIException::class)
    fun saveUser(user: User): User

    /**
     * Deactivate a user account so that it can no longer log in.
     *
     * @param user the user to retire
     * @param reason the reason for retirement
     * @return the retired user
     * @throws APIException if retirement fails
     */
    @Authorized(PrivilegeConstants.EDIT_USERS)
    @Throws(APIException::class)
    fun retireUser(user: User, reason: String): User

    /**
     * Clears retired flag for a user.
     *
     * @param user the user to unretire
     * @return the unretired user
     * @throws APIException if unretirement fails
     */
    @Authorized(PrivilegeConstants.EDIT_USERS)
    @Throws(APIException::class)
    fun unretireUser(user: User): User

    /**
     * Completely remove a user from the database (not reversible).
     *
     * @param user the User to remove from the database.
     * @throws APIException if purging fails
     */
    @Authorized(PrivilegeConstants.PURGE_USERS)
    @Throws(APIException::class)
    fun purgeUser(user: User)

    /**
     * Completely remove a user from the database (not reversible).
     *
     * @param user the User to remove from the database.
     * @param cascade true to delete associated content
     * @throws APIException if purging fails
     */
    @Authorized(PrivilegeConstants.PURGE_USERS)
    @Throws(APIException::class)
    fun purgeUser(user: User, cascade: Boolean)

    /**
     * Returns all privileges currently possible for any User.
     *
     * @return Global list of privileges
     * @throws APIException if retrieval fails
     */
    @Authorized(PrivilegeConstants.MANAGE_PRIVILEGES)
    @Throws(APIException::class)
    fun getAllPrivileges(): List<Privilege>

    /**
     * Returns all roles currently possible for any User.
     *
     * @return Global list of roles
     * @throws APIException if retrieval fails
     */
    @Authorized(PrivilegeConstants.MANAGE_ROLES)
    @Throws(APIException::class)
    fun getAllRoles(): List<Role>

    /**
     * Save the given role in the database.
     *
     * @param role Role to update
     * @return the saved role
     * @throws APIException if saving fails
     */
    @Authorized(PrivilegeConstants.MANAGE_ROLES)
    @Throws(APIException::class)
    fun saveRole(role: Role): Role

    /**
     * Completely remove a role from the database.
     *
     * @param role Role to delete from the database
     * @throws APIException if purging fails
     */
    @Authorized(PrivilegeConstants.PURGE_ROLES)
    @Throws(APIException::class)
    fun purgeRole(role: Role)

    /**
     * Save the given privilege in the database.
     *
     * @param privilege Privilege to update
     * @return the saved privilege
     * @throws APIException if saving fails
     */
    @Authorized(PrivilegeConstants.MANAGE_PRIVILEGES)
    @Throws(APIException::class)
    fun savePrivilege(privilege: Privilege): Privilege

    /**
     * Completely remove a privilege from the database.
     *
     * @param privilege Privilege to delete
     * @throws APIException if purging fails
     */
    @Authorized(PrivilegeConstants.PURGE_PRIVILEGES)
    @Throws(APIException::class)
    fun purgePrivilege(privilege: Privilege)

    /**
     * Returns role object with given string role.
     *
     * @param r the role name
     * @return Role object for specified string
     * @throws APIException if retrieval fails
     */
    @Authorized(PrivilegeConstants.GET_ROLES)
    @Throws(APIException::class)
    fun getRole(r: String): Role?

    /**
     * Get Role by its UUID.
     *
     * @param uuid the uuid
     * @return role or null
     * @throws APIException if retrieval fails
     */
    @Authorized(PrivilegeConstants.GET_ROLES)
    @Throws(APIException::class)
    fun getRoleByUuid(uuid: String): Role?

    /**
     * Returns Privilege in the system with given String privilege.
     *
     * @param p the privilege name
     * @return Privilege
     * @throws APIException if retrieval fails
     */
    @Authorized(PrivilegeConstants.GET_PRIVILEGES)
    @Throws(APIException::class)
    fun getPrivilege(p: String): Privilege?

    /**
     * Get Privilege by its UUID.
     *
     * @param uuid the uuid
     * @return privilege or null
     * @throws APIException if retrieval fails
     */
    @Authorized(PrivilegeConstants.GET_PRIVILEGES)
    @Throws(APIException::class)
    fun getPrivilegeByUuid(uuid: String): Privilege?

    /**
     * Returns all users in the system.
     *
     * @return Global list of users
     * @throws APIException if retrieval fails
     */
    @Authorized(PrivilegeConstants.GET_USERS)
    @Throws(APIException::class)
    fun getAllUsers(): List<User>

    /**
     * Changes the current user's password.
     *
     * @param oldPassword current password
     * @param newPassword new password
     * @throws APIException if change fails
     */
    @Authorized
    @Logging(ignoredArgumentIndexes = [0, 1])
    @Throws(APIException::class)
    fun changePassword(oldPassword: String, newPassword: String)

    /**
     * Changes password of User passed in.
     *
     * @param user user whose password is to be changed
     * @param newPassword new password to set
     * @throws APIException if change fails
     */
    @Authorized(PrivilegeConstants.EDIT_USER_PASSWORDS)
    @Throws(APIException::class)
    fun changePassword(user: User, newPassword: String)

    /**
     * Changes the current user's password directly.
     *
     * @param user the user whose password you want to change
     * @param hashedPassword - the already hashed password to store
     * @param salt - the salt which should be used with this hashed password
     * @throws APIException if change fails
     * @since 1.5
     */
    @Authorized(PrivilegeConstants.EDIT_USER_PASSWORDS)
    @Throws(APIException::class)
    fun changeHashedPassword(user: User, hashedPassword: String, salt: String)

    /**
     * Changes the passed user's secret question and answer.
     *
     * @param u User to change
     * @param question the question
     * @param answer the answer
     * @throws APIException if change fails
     * @since 1.5
     */
    @Authorized(PrivilegeConstants.EDIT_USER_PASSWORDS)
    @Logging(ignoredArgumentIndexes = [1, 2])
    @Throws(APIException::class)
    fun changeQuestionAnswer(u: User, question: String, answer: String)

    /**
     * Changes the current user's secret question and answer.
     *
     * @param pw user's password
     * @param q question
     * @param a answer
     * @throws APIException if change fails
     */
    @Authorized
    @Logging(ignoreAllArgumentValues = true)
    @Throws(APIException::class)
    fun changeQuestionAnswer(pw: String, q: String, a: String)

    /**
     * Returns secret question for the given user.
     *
     * @param user the user
     * @return the secret question
     * @throws APIException if retrieval fails
     * @since 2.0
     */
    @Throws(APIException::class)
    fun getSecretQuestion(user: User): String?

    /**
     * Compares answer against the user's secret answer.
     *
     * @param u user
     * @param answer the answer to check
     * @return true if the answer matches
     * @throws APIException if check fails
     */
    @Logging(ignoredArgumentIndexes = [1])
    @Throws(APIException::class)
    fun isSecretAnswer(u: User, answer: String): Boolean

    /**
     * Return a list of users sorted by personName if any part of the search matches
     * first/last/system id and the user has one at least one of the given roles assigned to them.
     *
     * @param nameSearch string to compare to the beginning of user's given/middle/family/family2 names
     * @param roles all the Roles the user must contain
     * @param includeVoided true/false whether to include voided users
     * @return list of users matching the given attributes
     * @throws APIException if retrieval fails
     */
    @Authorized(PrivilegeConstants.GET_USERS)
    @Throws(APIException::class)
    fun getUsers(nameSearch: String?, roles: @JvmSuppressWildcards List<Role>?, includeVoided: Boolean): List<User>

    /**
     * Search for a list of users by exact first name and last name.
     *
     * @param givenName the given name
     * @param familyName the family name
     * @param includeRetired whether to include retired users
     * @return List of users matching criteria
     * @throws APIException if retrieval fails
     */
    @Authorized(PrivilegeConstants.GET_USERS)
    @Throws(APIException::class)
    fun getUsersByName(givenName: String, familyName: String, includeRetired: Boolean): List<User>

    /**
     * Get all user accounts that belong to a given person.
     *
     * @param person the person
     * @param includeRetired whether to include retired accounts
     * @return all user accounts that belong to person
     * @throws APIException if retrieval fails
     */
    @Authorized(PrivilegeConstants.GET_USERS)
    @Throws(APIException::class)
    fun getUsersByPerson(person: Person, includeRetired: Boolean): List<User>

    /**
     * Adds the key/value pair to the given user.
     *
     * @param user the user
     * @param key the property key
     * @param value the property value
     * @return the user that was passed in and added to
     * @throws APIException if operation fails
     */
    @Authorized
    @Throws(APIException::class)
    fun setUserProperty(user: User, key: String, value: String): User?

    /**
     * Removes the property denoted by key from the user's properties.
     *
     * @param user the user
     * @param key the property key
     * @return the user that was passed in and removed from
     * @throws APIException if operation fails
     */
    @Authorized
    @Throws(APIException::class)
    fun removeUserProperty(user: User, key: String): User?

    /**
     * Get/generate/find the next system id to be doled out.
     *
     * @return new system id
     */
    @Authorized
    fun generateSystemId(): String

    /**
     * Return a batch of users of a specific size sorted by personName.
     *
     * @param name string to compare to the beginning of user's given/middle/family/family2 names
     * @param roles all the Roles the user must contain
     * @param includeRetired true/false whether to include voided users
     * @param start beginning index for the batch
     * @param length number of users to return in the batch
     * @return list of matching users of a size based on the specified arguments
     * @throws APIException if retrieval fails
     * @since 1.8
     */
    @Authorized(PrivilegeConstants.GET_USERS)
    @Throws(APIException::class)
    fun getUsers(name: String?, roles: @JvmSuppressWildcards List<Role>?, includeRetired: Boolean, start: Int?, length: Int?): List<User>

    /**
     * Return the number of users with a matching name or system id and have at least one of the
     * given roles assigned to them.
     *
     * @param name patient name
     * @param roles all the Roles the user must contain
     * @param includeRetired Specifies whether voided users should be included
     * @return the number of users matching the given attributes
     * @since 1.8
     */
    @Authorized(PrivilegeConstants.GET_USERS)
    fun getCountOfUsers(name: String?, roles: @JvmSuppressWildcards List<Role>?, includeRetired: Boolean): Int?

    /**
     * Saves the current key/value as a user property for the current user.
     *
     * @param key the authenticated user's property
     * @param value value of the property
     * @return the saved user
     * @since 1.10
     */
    @Authorized
    fun saveUserProperty(key: String, value: String): User

    /**
     * Replaces all user properties with the given map of properties for the current user.
     *
     * @param properties the authenticated user's properties
     * @return the saved user
     * @since 1.10
     */
    @Authorized
    fun saveUserProperties(properties: @JvmSuppressWildcards Map<String, String>): User

    /**
     * Change user password given the answer to the secret question.
     *
     * @param secretAnswer the answer to secret question
     * @param pw the new password
     * @throws APIException if change fails
     */
    @Authorized
    @Throws(APIException::class)
    fun changePasswordUsingSecretAnswer(secretAnswer: String, pw: String)

    /**
     * Sets a user's activation key.
     *
     * @param user The user for which the activation key will be set
     * @return the user with the activation key set
     * @throws MessageException if setting fails
     */
    @Authorized(PrivilegeConstants.EDIT_USER_PASSWORDS)
    @Throws(MessageException::class)
    fun setUserActivationKey(user: User): User

    /**
     * Change user password given the activation key.
     *
     * @param activationKey the activation for password reset
     * @param newPassword the new password
     */
    fun changePasswordUsingActivationKey(activationKey: String, newPassword: String)

    /**
     * Gets the default Locale of the given user.
     *
     * @param user the User whose Locale to retrieve
     * @return the default Locale of the given user, or the system locale if unspecified
     * @since 2.3.6, 2.4.6, 2.5.4, 2.6.0
     */
    @Authorized
    fun getDefaultLocaleForUser(user: User): Locale

    /**
     * Retrieves the last login time of the user in Unix Timestamp.
     *
     * @param user the subject user
     * @return timestamp representing last login time (e.g. 1717414410587)
     * @since 2.7.0
     */
    @Authorized
    fun getLastLoginTime(user: User): String
}
